src/main/java/com/hotels/bdp/circustrain/bigquery/client/BigQueryMetastoreClient.java

```diff
@@ -89,86 +93,124 @@
 import com.hotels.hcommon.hive.metastore.client.api.CloseableMetaStoreClient;
 
 class BigQueryMetastoreClient implements CloseableMetaStoreClient {
 
   private static final Logger log = LoggerFactory.getLogger(BigQueryMetastoreClient.class);
 
-  private final BigQuery bigQuery;
-  private final BigQueryDataExtractionManager dataExtractionManager;
+  private final CircusTrainBigQueryMetastore bigQueryMetastore;
+  private final ExtractionService extractionService;
+  private final TableServiceFactory tableServiceFactory;
+  private final MetastoreClientCache cache;
 
-  BigQueryMetastoreClient(BigQuery bigQuery, BigQueryDataExtractionManager dataExtractionManager) {
-    this.bigQuery = bigQuery;
-    this.dataExtractionManager = dataExtractionManager;
+  BigQueryMetastoreClient(
+      CircusTrainBigQueryMetastore bigQueryMetastore,
+      ExtractionService extractionService,
+      MetastoreClientCache metastoreClientCache,
+      TableServiceFactory tableServiceFactory) {
+    this.bigQueryMetastore = bigQueryMetastore;
+    this.extractionService = extractionService;
+    this.tableServiceFactory = tableServiceFactory;
+    this.cache = metastoreClientCache;
   }
 
-  private void checkDbExists(String databaseName) throws UnknownDBException {
-    if (bigQuery.getDataset(databaseName) == null) {
-      throw new UnknownDBException("Dataset " + databaseName + " doesn't exist in BigQuery");
-    }
+  @Override
+  public boolean isOpen() {
+    return true;
   }
 
+  @Override
+  public void close() {}
+
   @Override
   public Database getDatabase(String databaseName) throws TException {
-    log.info("Getting database {} from BigQuery", databaseName);
-    checkDbExists(databaseName);
+    log.info("Getting database '{}' from BigQuery", databaseName);
+    bigQueryMetastore.checkDbExists(databaseName);
     return new Database(databaseName, null, null, null);
   }
 
   @Override
   public boolean tableExists(String databaseName, String tableName) throws TException {
-    checkDbExists(databaseName);
-    return bigQuery.getDataset(databaseName).get(tableName) != null;
+    return bigQueryMetastore.tableExists(databaseName, tableName);
   }
 
   @Override
-  public Table getTable(String databaseName, String tableName) throws TException {
-    log.info("Getting table {}.{} from BigQuery", databaseName, tableName);
-    checkDbExists(databaseName);
-    com.google.cloud.bigquery.Table table = getBigQueryTable(databaseName, tableName);
-    Table hiveTable = new BigQueryToHiveTableConverter()
-        .withDatabaseName(databaseName)
-        .withTableName(tableName)
-        .withSchema(table.getDefinition().getSchema())
-        .withLocation(dataExtractionManager.location())
-        .convert();
+  public Table getTable(final String databaseName, final String tableName) throws TException {
+    log.info("Getting table '{}.{}' from BigQuery", databaseName, tableName);
+    String tableKey = makeKey(databaseName, tableName);
```

Comment:
Seeing as you're doing this key conversion a few times (here and on line 181 for example) and you don't need the key again in this method, why not add a method on the cache which takes the dbName and tableName as parameters and make the key there? Then this class won't need to know about the cache keys at all?